controllers/sample_local_helm_controller.go

```diff
@@ -0,0 +1,338 @@
+	case v1alpha1.StateProcessing:
+		return ctrl.Result{Requeue: true}, r.HandleProcessingState(ctx, &objectInstance)
+	case v1alpha1.StateDeleting:
+		return ctrl.Result{Requeue: true}, r.HandleDeletingState(ctx, &objectInstance)
+	case v1alpha1.StateError:
+		return ctrl.Result{Requeue: true}, r.HandleErrorState(ctx, &objectInstance)
+	case v1alpha1.StateReady:
+		return ctrl.Result{RequeueAfter: requeueInterval}, r.HandleReadyState(ctx, &objectInstance)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// HandleInitialState bootstraps state handling for the reconciled resource.
+func (r *SampleHelmReconciler) HandleInitialState(ctx context.Context, objectInstance *v1alpha1.SampleHelm) error {
+	status := objectInstance.Status
+
+	return r.setStatusForObjectInstance(ctx, objectInstance, status.
+		WithState(v1alpha1.StateProcessing).
+		WithInstallConditionStatus(metav1.ConditionUnknown, objectInstance.GetGeneration()))
+}
+
+// HandleProcessingState processes the reconciled resource by processing the underlying resources.
+// Based on the processing either a success or failure state is set on the reconciled resource.
+func (r *SampleHelmReconciler) HandleProcessingState(ctx context.Context, objectInstance *v1alpha1.SampleHelm) error {
+	status := objectInstance.Status
+	if err := r.processResources(ctx, objectInstance); err != nil {
+		r.Event(objectInstance, "Warning", "ResourcesInstall", err.Error())
+		return r.setStatusForObjectInstance(ctx, objectInstance, status.
+			WithState(v1alpha1.StateError).
+			WithInstallConditionStatus(metav1.ConditionFalse, objectInstance.GetGeneration()))
+	}
+	// set eventual state to Ready - if no errors were found
+	return r.setStatusForObjectInstance(ctx, objectInstance, status.
+		WithState(v1alpha1.StateReady).
+		WithInstallConditionStatus(metav1.ConditionTrue, objectInstance.GetGeneration()))
+}
+
+// HandleErrorState handles error recovery for the reconciled resource.
+func (r *SampleHelmReconciler) HandleErrorState(ctx context.Context, objectInstance *v1alpha1.SampleHelm) error {
+	status := objectInstance.Status
+	if err := r.processResources(ctx, objectInstance); err != nil {
+		return err
+	}
+	// set eventual state to Ready - if no errors were found
+	return r.setStatusForObjectInstance(ctx, objectInstance, status.
+		WithState(v1alpha1.StateReady).
+		WithInstallConditionStatus(metav1.ConditionTrue, objectInstance.GetGeneration()))
+}
+
+// HandleDeletingState processed the deletion on the reconciled resource.
+// Once the deletion if processed the relevant finalizers (if applied) are removed.
+func (r *SampleHelmReconciler) HandleDeletingState(ctx context.Context, objectInstance *v1alpha1.SampleHelm) error {
+	r.Event(objectInstance, "Normal", "Deleting", "resource deleting")
+	logger := log.FromContext(ctx)
+
+	status := objectInstance.Status
+
+	resources, err := r.render(ctx, objectInstance)
+	if err != nil && controllerutil.RemoveFinalizer(objectInstance, finalizer) {
```

Comment:
What if `controllerutil.RemoveFinalizer` returns false?

Reply:
if it returns false, there will be simply no update call anymore